### PR TITLE
New codechef scraper.

### PR DESCRIPTION
**Pull request by souravtecken**
1. Scraper takes care of all 3 codechef contests.
2. Adds easy points from divB to divA particpants automatically.
3. Takes care of different scoring systems for long, cookoff, lunchtime.

### CODE REVIEW

**Review by varunvora — COMMENTED** [review]
1. Is selenium required for this? Can we not scrape with BeautifulSoup? It is faster and easier to set up for a new person. We would need selenium only when we have to simulate clicks (like in Codejam)
2. What is the logic for combining Div A and Div B?
3. Did you test the output with other scraped codechef contests?

**Comment by souravtecken** [comment]
> 1. Is selenium required for this? Can we not scrape with BeautifulSoup? It is faster and easier to set up for a new person. We would need selenium only when we have to simulate clicks (like in Codejam)
> 2. What is the logic for combining Div A and Div B?
> 3. Did you test the output with other scraped codechef contests?

1. The scoreboard is obtained via an AJAX request, it's not part of the source, and they haven't specified their API url either. I assumed that beautiful soup couldn't be used to scrape since the scoreboard isn't part of the source.
2. I note the problem list in div B and div A, and find the number of problems in divB not in divA (set operation used). Then I find calculate total points to be awarded, for Long - 100, cookoff -1, so on.
Also, for cookoff, there's time involved too, so I convert the given time, let's say 1:23:34 to a number - 12345, and since it's in descending order of points, but ascending order of time, I subtract the time from 100,000 and add the points as the most significant digit on the left. Then I sort based on points (time based score generated for cookoff as stated). The same procedure for long and lunchtime, except there's no time factored in. 
3. I did, for several of them, that's how I found out that codechef updated ratings for the April contest. I shall do more tests though.

**Comment by varunvora** [comment]
> > 1. Is selenium required for this? Can we not scrape with BeautifulSoup? It is faster and easier to set up for a new person. We would need selenium only when we have to simulate clicks (like in Codejam)
> > 2. What is the logic for combining Div A and Div B?
> > 3. Did you test the output with other scraped codechef contests?
> 
> 1. The scoreboard is obtained via an AJAX request, it's not part of the source, and they haven't specified their API url either. I assumed that beautiful soup couldn't be used to scrape since the scoreboard isn't part of the source.
Try BeautifulSoup once and let me know. Because in general selenium is much slower and may be difficult to set up.
> 2. I note the problem list in div B and div A, and find the number of problems in divB not in divA (set operation used). Then I find calculate total points to be awarded, for Long - 100, cookoff -1, so on.
>    Also, for cookoff, there's time involved too, so I convert the given time, let's say 1:23:34 to a number - 12345, and since it's in descending order of points, but ascending order of time, I subtract the time from 100,000 and add the points as the most significant digit on the left. Then I sort based on points (time based score generated for cookoff as stated). The same procedure for long and lunchtime, except there's no time factored in.
Can you use the rank instead of relying on time? Because ultimately we just need the rank.
> 3. I did, for several of them, that's how I found out that codechef updated ratings for the April contest. I shall do more tests though.
Ok

**Comment by varunvora** [comment]
1. Try BeautifulSoup and let me know. I think the URL will have the resource and may not be using Ajax. Because in general selenium is much slower and may be difficult to set up.
2. Why don't we just use ranks instead of any time logic? Ultimately we need rank only.
3. Ok.

**Comment by souravtecken** [comment]
> 1. Try BeautifulSoup and let me know. I think the URL will have the resource and may not be using Ajax. Because in general selenium is much slower and may be difficult to set up.
> 2. Why don't we just use ranks instead of any time logic? Ultimately we need rank only.
> 3. Ok.

1. I tried doing what was done in hackerearth.py, the codechef url did not have the scoreboard in it. So, I don't think BeautifulSoup can be used for this purpse.
2. If we just use ranks, we can't place the divA particpants correctly in the scoreboard, for the Long contest, we had to manually add the scores and see where the divA particpants must be placed, but in the cookoff contest, the time's also a part of the score. A divA particpant with rank 2 must be placed above a divB participant with rank 1 in our scoreboard (considering both participants solved all problems of their division) according to the easy-point system, but we can't conclude that based on rank alone, so I've factored in score/time too.

**Comment by varunvora** [comment]
I still don't understand the need for considering time. In cookoff contests, time is also a factor but the rank depends on time. What if we gave the same rank to Div A and Div B contestant in case of ambiguity (instead of using time as a tiebreaker)?
A Div A may have spent more time on solving the harder problems.. we may not know. It'll be fair if we gave the same rank IMO.

**Comment by souravtecken** [comment]
> I still don't understand the need for considering time. In cookoff contests, time is also a factor but the rank depends on time. What if we gave the same rank to Div A and Div B contestant in case of ambiguity (instead of using time as a tiebreaker)?
> A Div A may have spent more time on solving the harder problems.. we may not know. It'll be fair if we gave the same rank IMO.

That's fair, I factored in time only for the tiebreaker. I probably should've suggested giving the same rank earlier, but I followed what they did there (not having multiple people with the same rank in cookoff). I shall remove the time part.

**Comment by souravtecken** [comment]
> I still don't understand the need for considering time. In cookoff contests, time is also a factor but the rank depends on time. What if we gave the same rank to Div A and Div B contestant in case of ambiguity (instead of using time as a tiebreaker)?
> A Div A may have spent more time on solving the harder problems.. we may not know. It'll be fair if we gave the same rank IMO.

If a Div A participant has solved the harder problems, they do get a better rank than the div B participants, because the score being calculated for each participant is a combination of time and number of problems solved. In this scoreboard for cookoff for example - [March Cook-off Div B](https://www.codechef.com/rankings/COOK104B), The top 5 people have 5 points, and yet they have different ranks, calculated by time. Now in case in a scoreboard we're considering, everybody's solved just 1 problem, if we go by points, we'd be assigning them all the same rank, which isn't fair. And if we go by ranks. Let's assume the best div B participant as solved 5 problems, and the best div A participant has solved 1 problem (the div B participant has also solved this problem). If we assign order rankwise, both these participants would be ranked together, which also isn't fair. So, I felt a point-time combination score would be fair. It does take care of the cases where a div A participant has solved a harder problem, because easy points are added to the score ( based on the assumption a div A participant can solve all easy div B problems ).

**Comment by varunvora** [comment]
Ok, what do you get when you run this script and compare with current rank files?

**Comment by souravtecken** [comment]
So far, this script has created rank files that are the same as earlier, but participation has been relatively low to come across any possible discrepancies that could occur.

**Comment by varunvora** [comment]
Same for all the codechef contests so far?

**Comment by souravtecken** [comment]
> Same for all the codechef contests so far?

Yeah

File: scrapers/codechef_selenium.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+scoreboard_base_url:str = "https://www.codechef.com/rankings"
+site_url="https://www.codechef.com"
+scoreboard_filter_query:str = "?filterBy=Institution%3DPES%20University%2C%20Bengaluru&order=asc&sortBy=rank"
+contest_code:str="APRIL19"
+name_class:str="user-name"
+
+divisions:list=["A","B"]
+problems:dict={'A':set(),'B':set()}
+scraped_scoreboard:dict={'A':list(),'B':list()}
```
Comment (varunvora) [inline scrapers/codechef_selenium.py]: There seem to be redundant objects with "A" and "B" as attributes. Is it possible to structure it? Something like `namedtuple` would be convenient.

File: scrapers/codechef_selenium.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    total_pages=int(driver.find_elements_by_class_name("jump")[-1].find_element_by_tag_name("a").text)
+    for page in range(total_pages):    
+        name_elements=driver.find_elements_by_class_name(name_class)
+        score_elements=driver.find_elements_by_xpath("//td[@class='num']//div[not(@class) and (@title='' or not(@title))]")
+        if contest_code[0:4]=="COOK":
+            score_elements=score_elements[::2]
+            time_elements=list(map(lambda x:100000-float(x.text.replace(":","")),driver.find_elements_by_class_name("total-time")))            
+            scraped_scoreboard[division].extend(zip(
+                [x.find_elements_by_tag_name("span")[-1].text for x in name_elements],
+                [float(score_elements[i].text.split()[0])*100000+time_elements[i] for i in range(len(score_elements))]))
```
Comment (varunvora) [inline scrapers/codechef_selenium.py:51]: Add comments for lines like these

**Review by varunvora — CHANGES_REQUESTED** [review]
Try to modularise the code and add comments at lines that are not readable.

**Review by varunvora — APPROVED** [review]

**Pull request merged**